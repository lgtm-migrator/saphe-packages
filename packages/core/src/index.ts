interface Test {
  hello: string;
}

export const testCore: Test = {
  hello: 'World',
};
