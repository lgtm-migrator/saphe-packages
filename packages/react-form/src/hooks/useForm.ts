import { useAsyncReducer, Util } from '@saphe/react-use';
import { useId, useMemo } from 'react';
import { Field, Fields } from '../lib/field';
import { FieldsBuilder, Plugins } from '../lib/plugin';
import { ComponentPropsBase, FieldPropsBase, PropsFromMany } from '../lib/props';
import { MaybePromise, OutputValue } from '../lib/util';

export interface FormConfig<P extends Plugins, F extends Fields> {
  /** Optional, declares the fields of the form */
  fields?: (t: FieldsBuilder<P>) => F;
  /** Optional, synchronous function that fires on a form field change event */
  onChange?: (
    formState: FormState<F>,
    targetValue: any,
    fieldName: keyof F,
    fieldIndex?: number,
  ) => FormState<F> | void;
  /** Optional, synchronous function that fires on a form field blur event */
  onBlur?: (
    formState: FormState<F>,
    fieldName: keyof F,
    fieldIndex?: number,
  ) => FormState<F> | void;
  /** Optional, sync or async function that fires after a form field change event */
  changeEffect?: (
    formState: FormState<F>,
    targetValue: any,
    fieldName: keyof F,
    fieldIndex?: number,
  ) => MaybePromise<void>;
  /** Optional, sync or async function that fires after a form field blur event */
  blurEffect?: (
    formState: FormState<F>,
    fieldName: keyof F,
    fieldIndex?: number,
  ) => MaybePromise<void>;
  /** Optional, defines what should happen on a submit event */
  onSubmit?: (formState: FormState<F>) => MaybePromise<FormState<F> | void>;
}

export type FormStateControls<F extends Fields> = {
  [I in keyof F]: F[I] extends Field<
    infer RawValue,
    infer _Value,
    infer Many,
    infer _Validation,
    infer _State,
    infer Options
  >
    ? PropsFromMany<RawValue, Many> & Options
    : never;
};

export type FormStateFields<F extends Fields> = {
  [I in keyof F]: F[I] extends Field<
    infer RawValue,
    infer Value,
    infer Many,
    infer Validation,
    infer State,
    infer Options
  >
    ? Field<RawValue, Value, Many, Validation, State, Options>
    : never;
};

export type FormStateValues<F extends Fields> = {
  [I in keyof F]: F[I] extends Field<
    infer _RawValue,
    infer Value,
    infer Many,
    infer Validation,
    infer _State,
    infer _Options
  >
    ? OutputValue<Value, Many, Validation>
    : never;
};

export type FormStateTouched<F extends Fields> = {
  [I in keyof F]: F[I] extends Field<
    infer _RawValue,
    infer _Value,
    infer Many,
    infer _Validation,
    infer _State,
    infer _Options
  >
    ? boolean extends Many
      ? boolean
      : Many extends true
      ? boolean[]
      : boolean
    : never;
};

export type FormStateError<F extends Fields> = {
  [I in keyof F]: F[I] extends Field<
    infer _RawValue,
    infer _Value,
    infer Many,
    infer _Validation,
    infer _State,
    infer _Options
  >
    ? boolean extends Many
      ? string
      : Many extends true
      ? string[]
      : string
    : never;
};

export type FormStateHelpers<_F extends Fields> = {
  formControl: {
    onSubmit: (e: { preventDefault: () => void }) => void;
    onReset: (e: { preventDefault: () => void }) => void;
  };
  submitButtonControl: {
    key: string;
    label: string;
    isDisabled: boolean;
    isLoading: boolean;
    type: 'submit' | 'reset' | 'button';
    onClick: (e: { preventDefault: () => void }) => void;
  };
};

export interface FormState<F extends Fields> {
  controls: FormStateControls<F>;
  fields: FormStateFields<F>;
  values: FormStateValues<F>;
  touched: FormStateTouched<F>;
  errors: FormStateError<F>;
  helpers: FormStateHelpers<F>;
}

export function useForm<P extends Plugins, F extends Fields>(
  plugins: P,
  config: FormConfig<P, F>,
): FormState<F> {
  const id = useId();

  const initialFields = useMemo<FormStateFields<F>>(
    () =>
      config.fields
        ? config.fields({
            field: Object.entries(plugins.fields).reduce(
              (prev, [pluginName, plugin]) => ({
                ...prev,
                [pluginName]: (opts: any) => ({ ...opts, plugin }),
              }),
              {} as any,
            ),
          })
        : ({} as any),
    [config.fields],
  );

  const initialControls = useMemo<FormStateControls<F>>(
    () =>
      Object.entries(initialFields).reduce((prev, [fieldName, field]) => {
        const componentProps: ComponentPropsBase = {
          label:
            field.label ??
            fieldName
              .replace(/([A-Z])/g, (match) => ` ${match}`)
              .replace(/^./, (match) => match.toUpperCase())
              .trim(),
          description: field.description ?? '',
        };

        const fieldProps = field.many
          ? (() => ({
              fields: ((field.initialValue as any[]) ?? [field.plugin.defaultInitialValue]).map(
                (value: any, index) => {
                  const props: FieldPropsBase<any> = {
                    id: `${id}${fieldName}${index}`,
                    name: `${id}${fieldName}${index}`,
                    describedBy: `${id}${fieldName}Description`,
                    value: field.plugin.serialize(value),
                    onChange: (targetValue: any) => actions.change(targetValue, fieldName, index),
                    onBlur: () => actions.blur(fieldName, index),

                    // TODO
                    isDisabled: false,
                    isHidden: false,
                    error: '',
                  };
                  return props;
                },
              ),
            }))()
          : (() => {
              const props: FieldPropsBase<any> = {
                id: `${id}${fieldName}`,
                name: `${id}${fieldName}`,
                describedBy: `${id}${fieldName}Description`,
                value: field.plugin.serialize(
                  field.initialValue ?? field.plugin.defaultInitialValue,
                ),
                onChange: (targetValue: any) => actions.change(targetValue, fieldName),
                onBlur: () => actions.blur(fieldName),

                // TODO
                isDisabled: false,
                isHidden: false,
                error: '',
              };
              return props;
            })();

        return {
          ...prev,
          [fieldName]: {
            ...field,
            ...componentProps,
            ...fieldProps,
          },
        };
      }, {} as FormStateControls<F>),
    [initialFields, id],
  );

  const initialValues = useMemo<FormStateValues<F>>(
    () =>
      Object.entries(initialFields).reduce(
        (prev, [fieldName, field]) => ({
          ...prev,
          [fieldName]: field.initialValue ?? field.plugin.defaultInitialValue,
        }),
        {} as FormStateValues<F>,
      ),
    [initialFields],
  );

  const initialTouched = useMemo<FormStateTouched<F>>(
    () =>
      Object.entries(initialFields).reduce(
        (prev, [fieldName, field]) => ({ ...prev, [fieldName]: field.many ? [false] : false }),
        {} as FormStateTouched<F>,
      ),
    [initialFields],
  );

  const initialErrors = useMemo<FormStateError<F>>(
    () =>
      Object.entries(initialFields).reduce(
        (prev, [fieldName, field]) => ({ ...prev, [fieldName]: field.many ? [''] : '' }),
        {} as FormStateError<F>,
      ),
    [initialFields],
  );

  // export function getInitialFormTouchedState<T extends Fields>(
  //   fields: T,
  //   prevTouched: Record<string, boolean> = {},
  // ): FormTouched<T> {
  //   const touched: Record<string, boolean> = {};
  //   for (const fieldName of Object.keys(fields)) touched[fieldName] = prevTouched[fieldName] ?? false;
  //   return touched as FormTouched<T>;
  // }

  // export function getInitialFormErrorsState<T extends Fields>(
  //   fields: T,
  //   prevErrors: Record<string, string> = {},
  // ): FormErrors<T> {
  //   const errors: Record<string, string> = {};
  //   for (const fieldName of Object.keys(fields)) errors[fieldName] = prevErrors[fieldName] ?? '';
  //   return errors as FormErrors<T>;
  // }

  // export function getInitialFormValuesState<T extends Fields>(
  //   fields: T,
  //   prevValues: Record<string, FieldValueCrude> = {},
  // ): FormValues<T> {
  //   const values: Record<string, FieldValueCrude> = {};
  //   for (const [fieldName, field] of Object.entries(fields)) {
  //     values[fieldName] = prevValues[fieldName] ?? field.initialValue ?? getDefaultFieldValue(field);
  //   }

  //   return values as FormValues<T>;
  // }

  // TODO helpers should be added afterwards, not included in state
  const initialHelpers = useMemo<FormStateHelpers<F>>(
    () => ({
      formControl: {
        onSubmit: (e) => {
          e.preventDefault();
          actions.submit();
        },
        onReset: (e) => {
          e.preventDefault();
          actions.reset();
        },
      },
      submitButtonControl: {
        key: `${id}submitButton`,
        type: 'submit',
        // TODO
        label: '',
        isDisabled: false,
        isLoading: false,
        // label: !isLoading
        //   ? submitButton?.label ?? 'Submit'
        //   : submitButton?.submittingLabel ?? submitButton?.label ?? 'Processing...',
        // disabled: Object.values(state.errors).reduce<boolean>(
        //   (prev, curr) => prev || curr !== '',
        //   false,
        // ),
        // isLoading,
        onClick: (e) => {
          e.preventDefault();
          actions.submit();
        },
      },
    }),
    [id],
  );

  const initialState: FormState<F> = useMemo(
    () => ({
      fields: initialFields,
      controls: initialControls,
      values: initialValues,
      touched: initialTouched,
      errors: initialErrors,
      helpers: initialHelpers,
    }),
    [initialFields, initialControls, initialValues, initialHelpers],
  );

  // TODO?
  // useEffect(() => {
  //   if (stringify(config.fields) !== stringify(state.fields)) actions.reset();
  // }, [actions, config.fields, state.fields]);

  // TODO isLoading, error
  const { state, actions } = useAsyncReducer(initialState, {
    reset: () => initialState,
    change: (state, targetValue: any, fieldName: keyof F, fieldIndex?: number) => {
      // TODO
      // const field = prevState.fields[fieldName];
      // if (!field) return prevState;

      // const values = {
      //   ...prevState.values,
      //   [fieldName]:
      //     'multiple' in field && field.multiple && targetValue === null
      //       ? []
      //       : 'multiple' in field && field.multiple && !(targetValue && Array.isArray(targetValue))
      //       ? [targetValue]
      //       : targetValue,
      // };

      // const error = validateField(field, targetValue);
      // const fieldValidationMode = field.validation?.mode;
      // const shouldValidate =
      //   // Either the local or global validation mode is ON_CHANGE
      //   fieldValidationMode === ValidationMode.ON_CHANGE ||
      //   (fieldValidationMode === undefined && validationMode === ValidationMode.ON_CHANGE) ||
      //   // Or either the local or global validation mode is AFTER_BLUR and a field has been touched
      //   ((fieldValidationMode === ValidationMode.AFTER_BLUR ||
      //     (fieldValidationMode === undefined && validationMode === ValidationMode.AFTER_BLUR)) &&
      //     prevState.touched[fieldName]);

      // return {
      //   ...prevState,
      //   values,
      //   errors: { ...prevState.errors, [fieldName]: shouldValidate ? error : '' },
      // };

      const field = state.fields[fieldName];

      const newValue = field.plugin.parse(targetValue);
      // if (field.many && newValue === null) newValue = [];
      // if (field.many && !Array.isArray(newValue)) newValue = [targetValue];

      let newState: FormState<F> = {
        ...state,
        controls: {
          ...state.controls,
          [fieldName]: {
            ...state.controls[fieldName],
            ...(fieldIndex === undefined
              ? { value: targetValue }
              : {
                  fields: Object.assign([], (state.controls[fieldName] as any).fields, {
                    [fieldIndex]: {
                      ...(state.controls[fieldName] as any).fields[fieldIndex],
                      value: targetValue,
                    },
                  }),
                }),
          },
        },
        values: {
          ...state.values,
          [fieldName]:
            fieldIndex === undefined
              ? newValue
              : Object.assign([], state.values[fieldName], { [fieldIndex]: newValue }),
        },
      };

      if (config.onChange) {
        const onChangeResult = config.onChange(newState, targetValue, fieldName, fieldIndex);
        if (typeof onChangeResult !== 'undefined') newState = onChangeResult;
      }

      if (config.changeEffect) {
        const changeEffectResult = config.changeEffect(
          newState,
          targetValue,
          fieldName,
          fieldIndex,
        );
        if (Util.isPromise(changeEffectResult)) Promise.resolve(changeEffectResult);
      }

      return newState;
    },
    blur: (state, fieldName: keyof F, fieldIndex?: number) => {
      // TODO
      // const field = prevState.fields[fieldName];
      // if (!field) return prevState;

      // const fieldValidationMode = field.validation?.mode;
      // const targetValue = prevState.values[fieldName as keyof typeof prevState.values];
      // const error = validateField(field, targetValue);

      // const shouldValidate =
      //   // Either the local or global validation mode is ON_BLUR
      //   fieldValidationMode === ValidationMode.ON_BLUR ||
      //   (fieldValidationMode === undefined && validationMode === ValidationMode.ON_BLUR) ||
      //   // Or either the local or global validation mode is AFTER_BLUR
      //   fieldValidationMode === ValidationMode.AFTER_BLUR ||
      //   (fieldValidationMode === undefined && validationMode === ValidationMode.AFTER_BLUR);

      // return {
      //   ...prevState,
      //   errors: { ...prevState.errors, [fieldName]: shouldValidate ? error : '' },
      //   touched: { ...prevState.touched, [fieldName]: true },
      // };

      let newState: FormState<F> = {
        ...state,
        touched: {
          ...state.touched,
          [fieldName]:
            fieldIndex === undefined
              ? true
              : Object.assign([], state.touched[fieldName], { [fieldIndex]: true }),
        },
      };

      if (config.onBlur) {
        const onChangeResult = config.onBlur(state, fieldName, fieldIndex);
        if (typeof onChangeResult !== 'undefined') newState = onChangeResult;
      }

      if (config.blurEffect) {
        const changeEffectResult = config.blurEffect(newState, fieldName, fieldIndex);
        if (Util.isPromise(changeEffectResult)) Promise.resolve(changeEffectResult);
      }

      return newState;
    },
    submit: async (state) => {
      let newState: FormState<F> = {
        ...state,
        touched: Object.entries(state.touched).reduce<FormStateTouched<F>>(
          (prev, [fieldName, touched]) => ({
            ...prev,
            [fieldName]: Array.isArray(touched) ? touched.map(() => true) : true,
          }),
          {} as FormStateTouched<F>,
        ),
      };

      let canSubmit = true;
      const newErrors: Record<string, string> = {};
      const newTouched: Record<string, boolean> = {};
      for (const [fieldName, _field] of Object.entries(newState.fields)) {
        // TODO
        // const error = validateField(field, state.values[fieldName as keyof typeof state.values]);
        const error = '';
        if (error !== '') canSubmit = false;
        newErrors[fieldName] = error;
        newTouched[fieldName] = true;
      }

      if (!canSubmit)
        return {
          ...newState,
          // TODO
          // errors: newErrors as FormErrors<T>,
          // touched: newTouched as FormTouched<T>,
        };

      if (config.onSubmit) {
        let onSubmitResult = config.onSubmit(newState);
        if (Util.isPromise(onSubmitResult)) onSubmitResult = await onSubmitResult;
        if (typeof onSubmitResult !== 'undefined') newState = onSubmitResult;
      }

      return newState;
    },
  });

  return state;
}
