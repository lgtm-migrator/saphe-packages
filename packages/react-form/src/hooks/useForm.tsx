import React, { useState, FormEvent, useEffect } from 'react';
import { Props } from '..';
import Form from '../components/Form';
import FieldContainer from '../components/helpers/FieldContainer';
import SubmitButton from '../components/helpers/SubmitButton';
import useRecaptcha from '../hooks/useRecaptcha';
import { FieldType } from '../lib/field';
import {
  getInitialFormErrorsState,
  getInitialFormTouchedState,
  getInitialFormValuesState,
} from '../lib/form';
import { Config, State } from '../lib/hook';
import { SubmitButtonProps } from '../lib/props';
import { Fields, FieldValue, FormErrors, FormTouched, FormValues, AddFieldPack } from '../lib/util';
import { validateField, ValidationMode } from '../lib/validation';

export default function useForm<T extends Fields>(config: Config<T>): State {
  const {
    name,
    fieldPack,
    validationMode = ValidationMode.AFTER_BLUR,
    recaptcha,
    submitButton,
    onSubmit,
  } = config;

  const [fields, setFields] = useState<T>(config.fields);

  // TODO rewrite with useReducer and extract handle functions?
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [touched, setTouched] = useState<FormTouched<T>>(getInitialFormTouchedState(config.fields));
  const [errors, setErrors] = useState<FormErrors<T>>(getInitialFormErrorsState(config.fields));
  const [values, setValues] = useState<FormValues<T>>(getInitialFormValuesState(config.fields));

  const { Recaptcha, recaptchaToken } = useRecaptcha(recaptcha);

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    // Form validation
    let canSubmit = true;
    const newErrors: Record<string, string> = {};
    const newTouched: Record<string, boolean> = {};
    for (const [fieldName, field] of Object.entries(fields)) {
      const error = validateField(field, values[fieldName]);
      if (error !== '') canSubmit = false;
      newErrors[fieldName] = error;
      newTouched[fieldName] = true;
    }
    if (!canSubmit) {
      setErrors(newErrors as FormErrors<T>);
      setTouched(newTouched as FormTouched<T>);
      return;
    }

    // Catch recaptcha errors
    if (recaptcha && !recaptchaToken) {
      recaptcha.onError();
      return;
    }

    // Fire custom submit function
    if (onSubmit) {
      setIsSubmitting(true);
      Promise.resolve(onSubmit(values, { recaptchaToken })).then(() => setIsSubmitting(false));
    }
  };

  const handleFieldChange = <T extends FieldType>(
    targetValue: FieldValue<T>,
    fieldName: string,
  ) => {
    let error = validateField(fields[fieldName], targetValue);
    const fieldValidationMode = fields[fieldName]?.validation?.mode;
    const shouldValidate =
      // Either the local or global validation mode is ON_CHANGE
      fieldValidationMode === ValidationMode.ON_CHANGE ||
      (fieldValidationMode === undefined && validationMode === ValidationMode.ON_CHANGE) ||
      // Or either the local or global validation mode is AFTER_BLUR and a field has been touched
      ((fieldValidationMode === ValidationMode.AFTER_BLUR ||
        (fieldValidationMode === undefined && validationMode === ValidationMode.AFTER_BLUR)) &&
        touched[fieldName]);

    if (!shouldValidate) error = '';

    setErrors({ ...errors, [fieldName]: error });
    setValues({ ...values, [fieldName]: targetValue });
  };

  const handleFieldBlur = (fieldName: string) => {
    let error = validateField(fields[fieldName], values[fieldName]);
    const fieldValidationMode = fields[fieldName]?.validation?.mode;
    const shouldValidate =
      fieldValidationMode === ValidationMode.ON_BLUR ||
      (fieldValidationMode === undefined && validationMode === ValidationMode.ON_BLUR) ||
      fieldValidationMode === ValidationMode.AFTER_BLUR ||
      (fieldValidationMode === undefined && validationMode === ValidationMode.AFTER_BLUR);

    if (!shouldValidate) error = '';

    setErrors({ ...errors, [fieldName]: error });
    setTouched({ ...touched, [fieldName]: true });
  };

  const submitButtonProps: AddFieldPack<Props.SubmitButtonProps> = {
    label: submitButton?.label ?? 'Submit',
    submittingLabel: submitButton?.submittingLabel ?? submitButton?.label ?? 'Submitting...',
    disabled: Object.values(errors).reduce<boolean>((prev, curr) => prev || curr !== '', false),
    fieldPack,
    isSubmitting,
  };

  useEffect(() => {
    if (JSON.stringify(config.fields) !== JSON.stringify(fields)) {
      setFields(config.fields);
      setTouched(getInitialFormTouchedState(config.fields, touched));
      setErrors(getInitialFormErrorsState(config.fields, errors));
      setValues(getInitialFormValuesState(config.fields, values));
    }
  }, [config.fields]);

  return {
    form: (
      <Form
        name={name}
        fields={fields}
        fieldPack={fieldPack}
        touched={touched}
        errors={errors}
        values={values}
        onSubmit={handleSubmit}
        onFieldBlur={handleFieldBlur}
        onFieldChange={handleFieldChange}
      >
        {recaptcha && (
          <FieldContainer fieldPack={fieldPack}>
            <Recaptcha />
          </FieldContainer>
        )}

        {!submitButton?.hidden && <SubmitButton {...submitButtonProps} />}
      </Form>
    ),
    submitButton: (
      <SubmitButton
        {...submitButtonProps}
        onClick={handleSubmit as never as SubmitButtonProps['onClick']}
      />
    ),
  };
}
