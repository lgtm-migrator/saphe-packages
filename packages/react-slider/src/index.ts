interface Test {
  hello: string;
}

const testSlider: Test = {
  hello: 'World'
}
