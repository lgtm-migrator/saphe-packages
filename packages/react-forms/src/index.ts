import { add } from '@saphe/core';

export const new_add = (a: number, b: number, c: number): number => {
  return add(add(a, b), c);
};
