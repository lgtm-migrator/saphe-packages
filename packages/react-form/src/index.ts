export { FieldTypes } from './utils/fieldTypes';
export { ValidationModes } from './utils/validationTypes';
export * as Props from './utils/propTypes';
export type { FieldPack } from './utils/helperTypes';
export type { Config } from './hooks/useForm';

import useForm from './hooks/useForm';
export default useForm;
