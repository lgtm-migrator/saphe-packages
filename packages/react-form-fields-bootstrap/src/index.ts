import { FieldPack, Field } from '@saphe/react-form';
import BSCheckField from './components/BSCheckField';
import BSFormFieldContainer from './components/BSFormFieldContainer';
import BSNumberField from './components/BSNumberField';
import BSSelectField from './components/BSSelectField';
import BSSubmitButton from './components/BSSubmitButton';
import BSTextAreaField from './components/BSTextAreaField';
import BSTextField from './components/BSTextField';

export const BootstrapFieldPack: FieldPack = {
  FormFieldContainer: BSFormFieldContainer,
  SubmitButton: BSSubmitButton,
  [Field.TEXT]: BSTextField,
  [Field.TEXT_AREA]: BSTextAreaField,
  [Field.SELECT]: BSSelectField,
  [Field.CHECK]: BSCheckField,
  [Field.NUMBER]: BSNumberField,
};
