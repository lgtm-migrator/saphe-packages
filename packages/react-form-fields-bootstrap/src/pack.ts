import { FieldPack, FieldTypes } from '@saphe/react-form';
import BSCheckBoxField from './components/BSCheckBoxField';
import BSSelectField from './components/BSSelectField';
import BSTextAreaField from './components/BSTextAreaField';
import BSTextField from './components/BSTextField';

// TODO add number input
export const BootstrapFieldPack: FieldPack = {
  [FieldTypes.TEXT]: BSTextField,
  [FieldTypes.TEXTAREA]: BSTextAreaField,
  [FieldTypes.SELECT]: BSSelectField,
  [FieldTypes.CHECKBOX]: BSCheckBoxField,
};
