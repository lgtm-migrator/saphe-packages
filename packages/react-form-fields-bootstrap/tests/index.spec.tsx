import useForm, { FieldTypes } from '@saphe/react-form';
import { renderHook } from '@testing-library/react-hooks';
import renderer from 'react-test-renderer';
import React from 'react';
import { BootstrapFieldPack } from '../src/index';
import { render } from '@testing-library/react';

describe('The field pack', () => {
  it('renders a minimal text field', () => {
    const {
      result: {
        current: { Form },
      },
    } = renderHook(() =>
      useForm({
        name: 'minimalTextFieldForm',
        fieldPack: BootstrapFieldPack,
        fields: {
          textField: {
            type: FieldTypes.TEXT,
            label: 'Text Field',
          },
        },
        onSubmit: () => console.log(),
      }),
    );
    
    expect(renderer.create(<Form />).toJSON()).toMatchSnapshot()
  });

  it('renders a minimal textarea field', () => {
    const {
      result: {
        current: { Form },
      },
    } = renderHook(() =>
      useForm({
        name: 'minimalTextareaFieldForm',
        fieldPack: BootstrapFieldPack,
        fields: {
          textField: {
            type: FieldTypes.TEXTAREA,
            label: 'Textarea Field',
          },
        },
        onSubmit: () => console.log(),
      }),
    );
    
    expect(renderer.create(<Form />).toJSON()).toMatchSnapshot()
  });

  it('renders a minimal select field', () => {
    const {
      result: {
        current: { Form },
      },
    } = renderHook(() =>
      useForm({
        name: 'minimalSelectFieldForm',
        fieldPack: BootstrapFieldPack,
        fields: {
          textField: {
            type: FieldTypes.SELECT,
            label: 'Select Field',
            options: [
              {
                label: 'Option 1',
                value: 'option-1',
              },
              {
                label: 'Option 2',
                value: 'option-2',
              },
            ],
          },
        },
        onSubmit: () => console.log(),
      }),
    );

    expect(renderer.create(<Form />).toJSON()).toMatchSnapshot()
  });

  it('renders a minimal checkbox field', () => {
    const {
      result: {
        current: { Form },
      },
    } = renderHook(() =>
      useForm({
        name: 'minimalCheckboxFieldForm',
        fieldPack: BootstrapFieldPack,
        fields: {
          textField: {
            type: FieldTypes.CHECKBOX,
            label: 'Checkbox Field',
          },
        },
        onSubmit: () => console.log(),
      }),
    );
    
    expect(renderer.create(<Form />).toJSON()).toMatchSnapshot()
  });
});
