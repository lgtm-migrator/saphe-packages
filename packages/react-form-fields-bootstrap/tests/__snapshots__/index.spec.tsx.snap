// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The field pack renders a minimal checkbox field 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="mb-3"
  >
    <div
      className="form-check"
    >
      <input
        aria-describedby="minimalCheckboxFieldFormTextFieldDescription"
        className="form-check-input"
        id="minimalCheckboxFieldFormTextField"
        name="textField"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="false"
      />
      <label
        className="form-check-label"
        htmlFor="minimalCheckboxFieldFormTextField"
      >
        Checkbox Field
      </label>
      
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`The field pack renders a minimal select field 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="mb-3"
  >
    <label
      className="form-label"
      htmlFor="minimalSelectFieldFormTextField"
    >
      Select Field
    </label>
    <select
      aria-describedby="minimalSelectFieldFormTextFieldDescription"
      className="form-select"
      id="minimalSelectFieldFormTextField"
      name="textField"
      onBlur={[Function]}
      onChange={[Function]}
      value="-placeholder-"
    >
      <option
        disabled={true}
        value="-placeholder-"
      >
        
      </option>
      <option
        value="option-1"
      >
        Option 1
      </option>
      <option
        value="option-2"
      >
        Option 2
      </option>
    </select>
    
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`The field pack renders a minimal text field 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="mb-3"
  >
    <label
      className="form-label"
      htmlFor="minimalTextFieldFormTextField"
    >
      Text Field
    </label>
    <input
      aria-describedby="minimalTextFieldFormTextFieldDescription"
      className="form-control"
      id="minimalTextFieldFormTextField"
      name="textField"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
    
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`The field pack renders a minimal textarea field 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="mb-3"
  >
    <label
      className="form-label"
      htmlFor="minimalTextareaFieldFormTextField"
    >
      Textarea Field
    </label>
    <textarea
      aria-describedby="minimalTextareaFieldFormTextFieldDescription"
      className="form-control"
      id="minimalTextareaFieldFormTextField"
      name="textField"
      onBlur={[Function]}
      onChange={[Function]}
      rows={6}
      value=""
    />
    
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;
